pkg/labels/parse.go

```diff
@@ -122,21 +126,55 @@ func ParseMatcher(s string) (_ *Matcher, err error) {
 	}
 
 	var (
-		rawValue            = ms[3]
-		value               strings.Builder
-		escaped             bool
-		expectTrailingQuote bool
+		rawName  = ms[1]
+		rawValue = ms[3]
 	)
 
-	if strings.HasPrefix(rawValue, "\"") {
-		rawValue = strings.TrimPrefix(rawValue, "\"")
-		expectTrailingQuote = true
+	matcherType, ok := typeMap[ms[2]]
+	if !ok {
+		return nil, errors.Errorf("matcher type is unknown: %s", ms[2])
 	}
 
+	if !utf8.ValidString(rawName) {
+		return nil, errors.Errorf("matcher label not valid UTF-8: %s", rawName)
+	}
 	if !utf8.ValidString(rawValue) {
-		return nil, errors.Errorf("matcher value not valid UTF-8: %s", ms[3])
+		return nil, errors.Errorf("matcher value not valid UTF-8: %s", rawValue)
+	}
+
+	value, err := unescapeMatcherString(rawValue)
+	if err != nil {
+		return nil, fmt.Errorf("matcher value %w", err)
+	}
+
+	name := rawName
+	// if name is quoted, then it can contain any UTF-8 character. Unescape some escape sequences.
+	if strings.HasPrefix(rawName, `"`) {
+		name, err = unescapeMatcherString(rawName)
+		if err != nil {
+			return nil, fmt.Errorf("matcher label name %w", err)
+		}
+	}
+
+	if !IsValidName(model.LabelName(name)) {
+		return nil, errors.Errorf("invalid matcher label name: \"%s\"", name)
 	}
 
+	return NewMatcher(matcherType, name, value)
+}
+
+// unescapeMatcherString unescapes sequences that are escaped by openMetricsEscape: \n, \\, \", and removes leading and trailing double quotes.
+func unescapeMatcherString(raw string) (string, error) {
+	var (
+		escaped             bool
+		value               strings.Builder
+		rawValue            = raw
+		expectTrailingQuote = false
+	)
+	if strings.HasPrefix(rawValue, `"`) {
```

Comment:
Should be removed before the call to `unescapeMatcherString` I think.

Reply:
What I mean here is that `unescapeMatcherString` is doing two operations: unescaping escaped sequences **and** also checking if a double quoted string has both start and end quotes. What I was proposing was separate those out into separate functions / code.